### ISSUE TITLE
Python3

### ISSUE THREAD

**Issue by Chris7** [Chris7/djangui#8]
Known issues for PY3 support:
Move argparse into its own PY3 compliant package
Fix bootstrapper

**Comment by Chris7** [comment]
PY3 complaint command line parser:
https://github.com/Chris7/clinto

**Comment by streetmapp** [comment]
Hey Chris,

Does this mean that Python3 support and support for other CLI packages (click) isn't far off?

**Comment by Chris7** [comment]
Yes it does, I'm writing some tox things right now to get some basic tests in for the parser.

**Comment by streetmapp** [comment]
That's awesome news. Since my first test case was both Python 3 and using Click. So seeing both djangui and wooey using Python 2.7 and argparse, I saw as another to do once I polished off the functionality. So I'll be ready to help with testing of that combination once you're ready for sure.

**Comment by Chris7** [comment]
Thanks, I'll make sure to let you know when I'm ready to move that code in. It's going to be Python3 before I start venturing out into all the other areas of CLIs, but I think they should be fairly redundant in what I need to parse out for compatibility.

**Comment by Chris7** [comment]
Preview port is available on the python3 branch

**Comment by streetmapp** [comment]
I've got a Djangui install up and running on Python 3.4.3. Looks like user auth on the main page may not be working. Can't login with my Admin user or test user, tells me credentials are wrong. Logging in with admin at /admin is fine though.

**Comment by streetmapp** [comment]
A few things I've come up with after some additional testing:
1. If using multiple positional arguments, it appears the order isn't what is displayed on the UI. For example if I have arguments arg1 and arg2, when I fill out the job form, they get passed arg2 arg1, causing the script to fail. 
2. When I try to run the script with the arguments swapped (to adjust for point 1), the job fails. This is the error being seen from the process screen:

```
21:52:45 worker.1 | [2015-06-02 01:52:45,909: INFO/MainProcess] Received task: djangui.tasks.submit_script[438ca3af-2704-49cf-8977-a18f74ec1936]
21:52:46 worker.1 | [2015-06-02 01:52:46,624: ERROR/MainProcess] Task djangui.tasks.submit_script[438ca3af-2704-49cf-8977-a18f74ec1936] raised unexpected: TypeError("Can't convert 'bytes' object to str implicitly",)
21:52:46 worker.1 | Traceback (most recent call last):
21:52:46 worker.1 |   File "/opt/python/envs/djangui/lib/python3.4/site-packages/celery/app/trace.py", line 240, in trace_task
21:52:46 worker.1 |     R = retval = fun(*args, **kwargs)
21:52:46 worker.1 |   File "/opt/python/envs/djangui/lib/python3.4/site-packages/celery/app/trace.py", line 438, in __protected_call__
21:52:46 worker.1 |     return self.run(*args, **kwargs)
21:52:46 worker.1 |   File "/opt/python/envs/djangui/lib/python3.4/site-packages/djangui/tasks.py", line 111, in submit_script
21:52:46 worker.1 |     utils.create_job_fileinfo(job)
21:52:46 worker.1 |   File "/opt/python/python34/lib/python3.4/contextlib.py", line 30, in inner
21:52:46 worker.1 |     return func(*args, **kwds)
21:52:46 worker.1 |   File "/opt/python/envs/djangui/lib/python3.4/site-packages/djangui/backend/utils.py", line 325, in create_job_fileinfo
21:52:46 worker.1 |     is_fasta, first_rows = test_fastx(filemodel['file'].path)
21:52:46 worker.1 |   File "/opt/python/envs/djangui/lib/python3.4/site-packages/djangui/backend/utils.py", line 311, in test_fastx
21:52:46 worker.1 |     seq += row
21:52:46 worker.1 | TypeError: Can't convert 'bytes' object to str implicitly

```

I can confirm the script works outside of djangui.

**Comment by Chris7** [comment]
Could you attach the script please?

Also -- the error is something outside of the script itself, it's in parsing the output of the file to tell which type it is. But it isn't PY3 compliant so thank you for catching that. I'll push updates for that issue shortly and check in the credentials.

**Comment by streetmapp** [comment]
Here you go nothing crazy, but will potentially be very useful for another team.

``` python
#!/opt/python/envs/djangui/bin/python

import argparse
import requests
from lxml import html


parser = argparse.ArgumentParser(description="idp admin cli")
parser.add_argument('link', help='the url containing the metadata')
parser.add_argument('name', help='the name of the file')

def getCert(url, filename):
    page = requests.get(url, verify=False)
    tree = html.fromstring(page.text)
    certs = tree.xpath('//x509certificate/text()')
    cert = []
    for i in certs:
        if i not in cert:
            cert.append(i)
    count = 1
    for a in cert:

        output_file = filename+str(count)+".crt"
        with open(output_file, 'w') as out:
            out.write("-----BEGIN CERTIFICATE-----\n")
            out.write(a+"\n")
            out.write("-----END CERTIFICATE-----")
        count += 1

if __name__ == '__main__':


    args = parser.parse_args()
    getCert(args.link,args.name)


```

**Comment by Chris7** [comment]
Thank you -- that was very helpful. I never use positional arguments in my own scripts so it's good to have these differences. This uncovered a few more bugs and forced me to work on script unit tests. It's updated now and should work with a fresh script import.

**Comment by Chris7** [comment]
user auth is fixed too. I for some reason was testing the jsonresponse for something I now forget and made it always fail for the python3 branch.

**Comment by streetmapp** [comment]
This is now working as expected. I am able to run my script and download the output (very slick having .zip and .tar.gz versions along with standalone files). 

I did come across another error when I logged out of my admin login from the app to test authentication, but am unable to reproduce that at this time. I'll post if it shows up again.

One thing that I started thinking about, is if there's potential to adjust how the output is dealt with. It's more so from thinking about cleaning up output to prevent the file system from filling up. I know I can change the default directory for this, but was thinking I may want to have a directory per script with no subdirectory per job run. It's a minor request in the scheme of things as I can script other things to clean up after X days or so, just something that popped into my head while using it.

Thanks for the quick turnarounds on the issues. Really great to get this moving.

**Comment by Chris7** [comment]
Great to hear. Thank you for reporting back. For the file output -- I have a plan in mind the is referenced in these 2 issues, #3 #4 . This will allow sharing of jobs/files between users, make choosing uploaded files easier, and management of user files. There are some existing tools that do it but unfortunately the ones I've seen are admin-based like django-filebrowser. If you have any experience with a django based file-browser that works nicely and doesn't come with a lot of dependencies I'd love to hear about it.

**Comment by Chris7** [comment]
Merged into master and released as djangui 0.2.6. Thanks for your feedback @streetmapp

**Comment by streetmapp** [comment]
No problem, glad to provide some insight into this. Thanks for creating this, and providing Python 3 support.
